google/cloud/bigtable/row_response.py

```diff
@@ -36,47 +37,284 @@ class RowResponse(Sequence["CellResponse"]):
     cells = row["family", "qualifier"]
     """
 
-    def __init__(self, key: row_key, cells: list[CellResponse]):
+    def __init__(
+        self,
+        key: row_key,
+        cells: list[CellResponse]
+        | dict[tuple[family_id, qualifier], list[dict[str, Any]]],
+    ):
+        """
+        Initializes a RowResponse object
+
+        RowResponse objects are not intended to be created by users.
+        They are returned by the Bigtable backend.
+        """
         self.row_key = key
-        self.cells: OrderedDict[
-            family_id, OrderedDict[qualifier, list[CellResponse]]
+        self._cells_map: dict[
+            family_id, dict[qualifier, list[CellResponse]]
         ] = OrderedDict()
-        """Expected to be used internally only"""
-        pass
+        self._cells_list: list[CellResponse] = []
+        if isinstance(cells, dict):
+            # handle dict input
+            tmp_list = []
+            for (family, qualifier), cell_list in cells.items():
+                for cell_dict in cell_list:
+                    cell_obj = CellResponse._from_dict(
+                        key, family, qualifier, cell_dict
+                    )
+                    tmp_list.append(cell_obj)
+            cells = tmp_list
+        # add cells to internal stores using Bigtable native ordering
+        for cell in sorted(cells):
+            if cell.row_key != self.row_key:
+                raise ValueError(
+                    f"CellResponse row_key ({cell.row_key!r}) does not match RowResponse key ({self.row_key!r})"
```

Comment:
Is this check necessary?

Reply:
Our backend should never create rows that hit this, but end users could

And it doesn't hurt to include it as a sanity check either way

Reply:
as discussed, endusers should never construct these